x/gamm/types/account.go

```diff
@@ -283,15 +278,89 @@ func (pa PoolAccount) GetAllPoolAssets() []PoolAsset {
+// TODO: (post-launch) If newWeights includes a new denomination,
+// add the balance as well to the pool's internal measurements.
+// TODO: (post-launch) If newWeights excludes an existing denomination,
+// remove the weight from the pool, and figure out something to do
+// with any remaining coin.
+func (pa PoolAccount) updateAllWeights(newWeights []PoolAsset) {
+	if len(pa.PoolAssets) != len(newWeights) {
+		panic("updateAllWeights called with invalid input, len(newWeights) != len(existingWeights)")
+	}
+	for i, asset := range pa.PoolAssets {
+		if asset.Token.Denom != newWeights[i].Token.Denom {
+			panic(fmt.Sprintf("updateAllWeights called with invalid input, "+
+				"expected new weights' %vth asset to be %v, got %v",
+				i, asset.Token.Denom, newWeights[i].Token.Denom))
+		}
+		err := newWeights[i].ValidateWeight()
+		if err != nil {
+			panic("updateAllWeights: Tried to set an invalid weight")
+		}
+		pa.PoolAssets[i].Weight = newWeights[i].Weight
 	}
+}
 
-	PoolAsset.Weight = weight
+// PokeTokenWeights checks to see if the pool's token weights need to be updated,
+// and if so, does so.
+func (pa PoolAccount) PokeTokenWeights(blockTime time.Time) {
+	// Pool weights aren't changing, do nothing.
+	poolWeightsChanging := (pa.PoolParams.SmoothWeightChangeParams != nil)
+	if !poolWeightsChanging {
+		return
+	}
+	// Pool weights are changing.
+	// TODO: Add intra-block cache check that we haven't already poked
+	// the block yet.
+	params := *pa.PoolParams.SmoothWeightChangeParams
+	// the weights w(t) for the pool at time `t` is the following:
+	//   t <= start_time: w(t) = initial_pool_weights
+	//   start_time < t <= start_time + duration:
+	//     w(t) = initial_pool_weights + (t - start_time) *
+	//       (target_pool_weights - initial_pool_weights) / (duration)
+	//   t > start_time + duration: w(t) = target_pool_weights
+
+	// t <= StartTime
+	if blockTime.Before(params.StartTime) || params.StartTime.Equal(blockTime) {
+		// Do nothing
+		return
+	} else if blockTime.After(params.StartTime.Add(params.Duration)) {
+		// t > start_time + duration
+		// Update weights to be the target weights.
+		// TODO: When we add support for adding new assets via this method,
+		// 		 Ensure the new asset has some token sent with it.
+		pa.updateAllWeights(params.TargetPoolWeights)
+		// We've finished updating weights, so delete this parameter
+		// TODO: This line doesn't work, since this is a non-pointer receiever,
+		// and pa.PoolParams gets copied.
+		pa.PoolParams.SmoothWeightChangeParams = nil
+		return
+	} else {
+		//	w(t) = initial_pool_weights + (t - start_time) *
+		//       (target_pool_weights - initial_pool_weights) / (duration)
+		// We first compute percent duration elapsed = (t - start_time) / duration, via Unix time.
+		shiftedBlockTime := blockTime.Sub(params.StartTime).Milliseconds()
+		percent_duration_elapsed := sdk.NewDec(shiftedBlockTime).QuoInt64(params.Duration.Milliseconds())
```

Comment:
This is fairly minor, but this being the only time snake case feels a bit strange.